File: snippets/screens/compose/onboarding_carousel_screen/3/onboarding_carousel_screen.kt
package com.aiq.screens

import androidx.compose.animation.core.*
import androidx.compose.foundation.ExperimentalFoundationApi
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.pager.HorizontalPager
import androidx.compose.foundation.pager.rememberPagerState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import kotlinx.coroutines.launch

data class OnboardingPage(
    val icon: ImageVector,
    val title: String,
    val description: String,
    val color: Color
)

@OptIn(ExperimentalFoundationApi::class)
@Composable
fun OnboardingCarouselScreen(
    onNavigateToWelcome: () -> Unit = {},
    onSkip: () -> Unit = {}
) {
    val pages = listOf(
        OnboardingPage(
            icon = Icons.Default.Psychology,
            title = "Discover Your AI Potential",
            description = "Explore the fascinating world of artificial intelligence and discover how smart you really are.",
            color = Color(0xFF6366F1)
        ),
        OnboardingPage(
            icon = Icons.Default.Quiz,
            title = "Take Smart Challenges",
            description = "Engage with carefully crafted questions designed to test your AI knowledge and reasoning skills.",
            color = Color(0xFF8B5CF6)
        ),
        OnboardingPage(
            icon = Icons.Default.TrendingUp,
            title = "Track Your Progress",
            description = "Monitor your improvement over time and compete with friends to see who has the highest AIQ score.",
            color = Color(0xFF06B6D4)
        )
    )

    val pagerState = rememberPagerState(pageCount = { pages.size })
    val coroutineScope = rememberCoroutineScope()

    // Animate background color based on current page
    val currentPage = pagerState.currentPage
    val backgroundColorAnim by animateColorAsState(
        targetValue = pages[currentPage].color,
        animationSpec = tween(500)
    )

    Box(
        modifier = Modifier
            .fillMaxSize()
            .background(
                brush = Brush.verticalGradient(
                    colors = listOf(
                        backgroundColorAnim,
                        backgroundColorAnim.copy(alpha = 0.8f)
                    )
                )
            )
    ) {
        Column(
            modifier = Modifier.fillMaxSize()
        ) {
            // Skip button
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(16.dp),
                horizontalArrangement = Arrangement.End
            ) {
                TextButton(onClick = onSkip) {
                    Text(
                        text = "Skip",
                        color = Color.White,
                        fontSize = 16.sp,
                        fontWeight = FontWeight.Medium
                    )
                }
            }

            // Page content
            HorizontalPager(
                state = pagerState,
                modifier = Modifier
                    .fillMaxWidth()
                    .weight(1f)
            ) { page ->
                OnboardingPageContent(
                    page = pages[page],
                    modifier = Modifier
                        .fillMaxSize()
                        .padding(24.dp)
                )
            }

            // Page indicators and navigation
            Column(
                modifier = Modifier.padding(24.dp),
                horizontalAlignment = Alignment.CenterHorizontally
            ) {
                // Page indicators
                Row(
                    horizontalArrangement = Arrangement.spacedBy(4.dp)
                ) {
                    repeat(pages.size) { index ->
                        val width by animateDpAsState(
                            targetValue = if (currentPage == index) 24.dp else 8.dp,
                            animationSpec = tween(300)
                        )

                        Box(
                            modifier = Modifier
                                .height(8.dp)
                                .width(width)
                                .clip(RoundedCornerShape(4.dp))
                                .background(
                                    color = if (currentPage == index)
                                        Color.White
                                    else
                                        Color.White.copy(alpha = 0.4f)
                                )
                        )
                    }
                }

                Spacer(modifier = Modifier.height(32.dp))

                // Next button
                Button(
                    onClick = {
                        if (currentPage < pages.size - 1) {
                            coroutineScope.launch {
                                pagerState.animateScrollToPage(currentPage + 1)
                            }
                        } else {
                            onNavigateToWelcome()
                        }
                    },
                    modifier = Modifier
                        .fillMaxWidth()
                        .height(56.dp),
                    colors = ButtonDefaults.buttonColors(
                        containerColor = Color.White,
                        contentColor = pages[currentPage].color
                    ),
                    shape = RoundedCornerShape(16.dp)
                ) {
                    Text(
                        text = if (currentPage == pages.size - 1) "Get Started" else "Next",
                        fontSize = 16.sp,
                        fontWeight = FontWeight.SemiBold
                    )
                    if (currentPage < pages.size - 1) {
                        Spacer(modifier = Modifier.width(8.dp))
                        Icon(
                            imageVector = Icons.Default.ArrowForward,
                            contentDescription = null,
                            modifier = Modifier.size(20.dp)
                        )
                    }
                }
            }
        }
    }
}

@Composable
fun OnboardingPageContent(
    page: OnboardingPage,
    modifier: Modifier = Modifier
) {
    Column(
        modifier = modifier,
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.Center
    ) {
        // Icon container
        Card(
            modifier = Modifier.size(120.dp),
            shape = CircleShape,
            colors = CardDefaults.cardColors(
                containerColor = Color.White.copy(alpha = 0.2f)
            )
        ) {
            Box(
                modifier = Modifier.fillMaxSize(),
                contentAlignment = Alignment.Center
            ) {
                Icon(
                    imageVector = page.icon,
                    contentDescription = null,
                    tint = Color.White,
                    modifier = Modifier.size(60.dp)
                )
            }
        }

        Spacer(modifier = Modifier.height(48.dp))

        // Title
        Text(
            text = page.title,
            fontSize = 28.sp,
            fontWeight = FontWeight.Bold,
            color = Color.White,
            textAlign = TextAlign.Center,
            lineHeight = 34.sp
        )

        Spacer(modifier = Modifier.height(24.dp))

        // Description
        Text(
            text = page.description,
            fontSize = 16.sp,
            color = Color.White.copy(alpha = 0.7f),
            textAlign = TextAlign.Center,
            lineHeight = 24.sp
        )
    }
}

@Preview(showBackground = true)
@Composable
fun OnboardingCarouselScreenPreview() {
    OnboardingCarouselScreen()
}

File: snippets/screens/swiftui/onboarding_carousel_screen/2/onboarding_carousel_screen.swift
import SwiftUI

struct OnboardingCarouselScreen: View {
    @State private var currentIndex = 0
    @State private var dragOffset: CGSize = .zero

    private let onboardingPages: [OnboardingPage] = [
        OnboardingPage(
            icon: "brain.head.profile",
            title: "Discover Your AI Potential",
            description: "Explore the fascinating world of artificial intelligence and discover how smart you really are.",
            color: Color(red: 0.39, green: 0.40, blue: 0.95)
        ),
        OnboardingPage(
            icon: "questionmark.circle.fill",
            title: "Take Smart Challenges",
            description: "Engage with carefully crafted questions designed to test your AI knowledge and reasoning skills.",
            color: Color(red: 0.55, green: 0.36, blue: 0.96)
        ),
        OnboardingPage(
            icon: "chart.line.uptrend.xyaxis",
            title: "Track Your Progress",
            description: "Monitor your improvement over time and compete with friends to see who has the highest AIQ score.",
            color: Color(red: 0.02, green: 0.71, blue: 0.83)
        )
    ]

    var body: some View {
        GeometryReader { geometry in
            ZStack {
                // Background gradient
                LinearGradient(
                    gradient: Gradient(colors: [
                        onboardingPages[currentIndex].color,
                        onboardingPages[currentIndex].color.opacity(0.8)
                    ]),
                    startPoint: .top,
                    endPoint: .bottom
                )
                .ignoresSafeArea()
                .animation(.easeInOut(duration: 0.5), value: currentIndex)

                VStack {
                    // Skip button
                    HStack {
                        Spacer()
                        Button("Skip") {
                            navigateToWelcome()
                        }
                        .foregroundColor(.white)
                        .font(.system(size: 16, weight: .medium))
                        .padding()
                    }

                    Spacer()

                    // Page content
                    TabView(selection: $currentIndex) {
                        ForEach(Array(onboardingPages.enumerated()), id: \.offset) { index, page in
                            OnboardingPageView(page: page)
                                .tag(index)
                        }
                    }
                    .tabViewStyle(PageTabViewStyle(indexDisplayMode: .never))
                    .frame(height: geometry.size.height * 0.6)

                    Spacer()

                    // Page indicators and navigation
                    VStack(spacing: 32) {
                        // Page indicators
                        HStack(spacing: 8) {
                            ForEach(0..<onboardingPages.count, id: \.self) { index in
                                RoundedRectangle(cornerRadius: 4)
                                    .fill(currentIndex == index ? Color.white : Color.white.opacity(0.4))
                                    .frame(width: currentIndex == index ? 24 : 8, height: 8)
                                    .animation(.easeInOut(duration: 0.3), value: currentIndex)
                            }
                        }

                        // Next button
                        Button(action: nextPage) {
                            HStack {
                                Text(currentIndex == onboardingPages.count - 1 ? "Get Started" : "Next")
                                    .font(.system(size: 16, weight: .semibold))
                                if currentIndex < onboardingPages.count - 1 {
                                    Image(systemName: "arrow.right")
                                        .font(.system(size: 14, weight: .semibold))
                                }
                            }
                            .foregroundColor(onboardingPages[currentIndex].color)
                            .frame(maxWidth: .infinity)
                            .frame(height: 56)
                            .background(Color.white)
                            .cornerRadius(16)
                        }
                    }
                    .padding(.horizontal, 24)
                    .padding(.bottom, 24)
                }
            }
        }
        .preferredColorScheme(.dark)
    }

    private func nextPage() {
        if currentIndex < onboardingPages.count - 1 {
            withAnimation(.easeInOut(duration: 0.3)) {
                currentIndex += 1
            }
        } else {
            navigateToWelcome()
        }
    }

    private func navigateToWelcome() {
        // Navigate to welcome screen
        // NavigationManager.shared.navigate(to: .welcome)
    }
}

struct OnboardingPageView: View {
    let page: OnboardingPage

    var body: some View {
        VStack(spacing: 48) {
            // Icon
            ZStack {
                Circle()
                    .fill(Color.white.opacity(0.2))
                    .frame(width: 120, height: 120)

                Image(systemName: page.icon)
                    .font(.system(size: 60))
                    .foregroundColor(.white)
            }

            // Content
            VStack(spacing: 24) {
                Text(page.title)
                    .font(.system(size: 28, weight: .bold))
                    .foregroundColor(.white)
                    .multilineTextAlignment(.center)
                    .lineLimit(nil)

                Text(page.description)
                    .font(.system(size: 16))
                    .foregroundColor(.white.opacity(0.7))
                    .multilineTextAlignment(.center)
                    .lineLimit(nil)
                    .lineSpacing(4)
            }
            .padding(.horizontal, 24)
        }
    }
}

struct OnboardingPage {
    let icon: String
    let title: String
    let description: String
    let color: Color
}

#Preview {
    OnboardingCarouselScreen()
}